File: grammar/ts/5_exercises.ts
// TypeScript 语言练习题
// 请完成以下练习题，每个函数都有详细的说明和示例
// 重点练习 TypeScript 的类型系统、接口、泛型、类等核心概念

// ==================== 练习题 ====================

// 练习1: 基础类型和类型注解
// 完成函数，计算两个数字的和、差、积、商
interface CalculationResult {
    sum: number;
    difference: number;
    product: number;
    quotient: number;
}

function calculate(a: number, b: number): CalculationResult {
    // TODO: 实现计算逻辑
    // 返回包含四种运算结果的对象
    return {
        sum: a + b,
        difference: a - b,
        product: a * b,
        quotient: a / b
    }; 
}

// 练习2: 字符串操作和联合类型
type StringOperation = "uppercase" | "lowercase" | "capitalize" | "reverse";

function processString(input: string, operation: StringOperation): string {
    // TODO: 根据操作类型处理字符串
    switch (operation) {
        case "uppercase":
            return input.toUpperCase();
        case "lowercase":
            return input.toLowerCase();
        case "capitalize":
            return input.charAt(0).toUpperCase() + input.slice(1);
        case "reverse":
            return input.split("").reverse().join("");
        default:
            throw new Error("Invalid operation");
    }
}

// 练习3: 数组操作和泛型
// 完成泛型函数，找出数组中的最大值和最小值
interface MinMax<T> {
    min: T;
    max: T;
}

function findMinMax<T>(array: T[], compareFn: (a: T, b: T) => number): MinMax<T> | null {
    // TODO: 实现找出最大值和最小值的逻辑
    // 如果数组为空，返回 null
    // 使用 compareFn 进行比较
    if (array.length === 0) {
        return null;
    }
    let min = array[0];
    let max = array[0];
    for (let i = 1; i < array.length; i++) {
        if (compareFn(array[i], min) < 0) {
            min = array[i];
        }
        if (compareFn(array[i], max) > 0) {
            max = array[i];
        }
    }
    return { min, max };
}

// 练习4: 接口和类
// 定义一个学生管理系统
interface Student {
    // TODO: 定义学生接口
    id: number;
    name: string;
    age: number;
    grades: number[];
    getAverageGrade(): number;
    addGrade(grade: number): void;
}

class StudentClass implements Student {
    // TODO: 实现 Student 接口
    id: number = 0;
    name: string = "";
    age: number = 0;
    grades: number[] = [];

    constructor(id: number, name: string, age: number) {
        // TODO: 初始化属性
        this.id = id;
        this.name = name;
        this.age = age;
    }

    getAverageGrade(): number {
        // TODO: 计算平均成绩
        if (this.grades.length === 0) {
            return 0;
        }
        const sum = this.grades.reduce((acc, cur) => acc + cur, 0);
        return sum / this.grades.length;
    }

    addGrade(grade: number): void {
        // TODO: 添加成绩到数组
        this.grades.push(grade);
    }

    // TODO: 添加一个方法获取学生信息
    getInfo(): string {
        return `ID: ${this.id}, Name: ${this.name}, Age: ${this.age}, Grades: ${this.grades}`;
    }
}

// 练习5: 枚举和类型守卫
enum TaskStatus {
    // TODO: 定义任务状态枚举
    PENDING = "pending",
    IN_PROGRESS = "in_progress",
    COMPLETED = "completed",
    CANCELLED = "cancelled"
}

interface Task {
    id: number;
    title: string;
    status: TaskStatus;
    createdAt: Date;
    completedAt?: Date;
}

// 类型守卫函数
function isCompleted(task: Task): task is Task & { completedAt: Date } {
    // TODO: 实现类型守卫，检查任务是否已完成
    return task.status === TaskStatus.COMPLETED;
}

function updateTaskStatus(task: Task, newStatus: TaskStatus): Task {
    // TODO: 更新任务状态
    // 如果状态变为 COMPLETED，设置 completedAt
    // 如果状态从 COMPLETED 变为其他状态，清除 completedAt
    if (newStatus !== TaskStatus.COMPLETED && isCompleted(task)) {
        task.completedAt = new Date();
    }
    task.status = newStatus;
    return task;
}

// 练习6: 泛型约束和映射类型
// 定义一个泛型函数，只接受有 length 属性的类型
interface Lengthwise {
    length: number;
}

function logLength<T extends Lengthwise>(arg: T): T {
    // TODO: 打印参数的长度并返回参数
    console.log(`Length: ${arg.length}`);
    return arg;
}

// 映射类型练习：创建一个类型，将所有属性变为可选
type Partial<T> = {
    // TODO: 实现 Partial 映射类型
    // [P in keyof T]?: T[P];
    [P in keyof T]?: T[P];
};

// 练习7: 函数重载
// 实现一个格式化函数，支持不同类型的输入
function format(value: string): string;
function format(value: number): string;
function format(value: Date): string;
function format(value: boolean): string;
function format(value: string | number | Date | boolean): string {
    // TODO: 根据不同类型实现格式化逻辑
    // string: 添加引号
    // number: 添加千位分隔符
    // Date: 格式化为 YYYY-MM-DD
    // boolean: 转换为 Yes/No
    if (typeof value === "string") {
        return `"${value}"`;
    } else if (typeof value === "number") {
        return value.toLocaleString();
    } else if (value instanceof Date) {
        return value.toISOString().split("T")[0];
    } else if (typeof value === "boolean") {
        return value ? "Yes" : "No";
    }
    return "";
}

// 练习8: 装饰器（实验性功能）
// 创建一个简单的日志装饰器
function log(target: any, propertyName: string, descriptor: PropertyDescriptor) {
    // TODO: 实现日志装饰器
    // 在方法执行前后打印日志
    const method = descriptor.value;
    
    descriptor.value = function (...args: any[]) {
        // TODO: 添加日志逻辑
        console.log(`调用方法: ${propertyName}`);
        const result = method.apply(this, args);
        console.log(`方法 ${propertyName} 执行完成`);
        return result;
    };
}

class Calculator {
    @log
    add(a: number, b: number): number {
        return a + b;
    }

    @log
    multiply(a: number, b: number): number {
        return a * b;
    }
}

// 练习9: Promise 和异步操作
// 实现一个异步数据获取函数
interface ApiResponse<T> {
    data: T;
    status: number;
    message: string;
}

async function fetchData<T>(url: string): Promise<ApiResponse<T>> {
    // TODO: 模拟异步数据获取
    // 随机延迟 1-3 秒
    // 90% 概率成功，10% 概率失败
    await new Promise(resolve => setTimeout(resolve, Math.random() * 2000 + 1000));
    // 90% 概率成功，10% 概率失败
    if (Math.random() < 0.9) {
        return {
            data: {} as T,
            status: 200,
            message: "Success"
        };
    } else {
        throw new Error("模拟 API 调用失败");
    }
}

// 练习10: 高级类型操作
// 实现一个深度只读类型
type DeepReadonly<T> = {
    // TODO: 实现深度只读类型
    // readonly [P in keyof T]: T[P] extends object ? DeepReadonly<T[P]> : T[P];
    readonly [P in keyof T]: T[P] extends object ? DeepReadonly<T[P]> : T[P];
};

// 实现一个选择特定属性的类型
type Pick<T, K extends keyof T> = {
    // TODO: 实现 Pick 类型
    // [P in K]: T[P];
    [P in K]: T[P];
};

// ==================== 测试函数 ====================

function testCalculate(): void {
    console.log("=== 测试计算函数 ===");
    const result = calculate(10, 3);
    console.log("10 和 3 的计算结果:", result);
    // 期望结果: { sum: 13, difference: 7, product: 30, quotient: 3.33... }
}

function testStringProcessing(): void {
    console.log("\n=== 测试字符串处理 ===");
    const text = "hello world";
    console.log(`原文: ${text}`);
    console.log(`大写: ${processString(text, "uppercase")}`);
    console.log(`小写: ${processString(text, "lowercase")}`);
    console.log(`首字母大写: ${processString(text, "capitalize")}`);
    console.log(`反转: ${processString(text, "reverse")}`);
}

function testMinMax(): void {
    console.log("\n=== 测试最大最小值 ===");
    const numbers = [3, 7, 2, 9, 1, 5, 8];
    const result = findMinMax(numbers, (a, b) => a - b);
    console.log(`数组 [${numbers.join(", ")}] 的最大最小值:`, result);
    
    const strings = ["apple", "banana", "cherry", "date"];
    const stringResult = findMinMax(strings, (a, b) => a.localeCompare(b));
    console.log(`字符串数组的最大最小值:`, stringResult);
}

function testStudent(): void {
    console.log("\n=== 测试学生类 ===");
    // TODO: 创建学生实例并测试方法
    const student = new StudentClass(1, "张三", 20);
    student.addGrade(85);
    student.addGrade(92);
    student.addGrade(78);
    console.log(student.getInfo());
    console.log(`平均成绩: ${student.getAverageGrade()}`);
}

function testTaskManagement(): void {
    console.log("\n=== 测试任务管理 ===");
    // TODO: 创建任务并测试状态更新
    const task: Task = {
        id: 1,
        title: "完成 TypeScript 练习",
        status: TaskStatus.PENDING,
        createdAt: new Date()
    };
    
    console.log("初始任务:", task);
    const updatedTask = updateTaskStatus(task, TaskStatus.COMPLETED);
    console.log("更新后任务:", updatedTask);
    console.log("任务是否完成:", isCompleted(updatedTask));
}

function testGenericConstraints(): void {
    console.log("\n=== 测试泛型约束 ===");
    logLength("Hello World");
    logLength([1, 2, 3, 4, 5]);
    logLength({ length: 10, value: "test" });
}

function testFunctionOverloads(): void {
    console.log("\n=== 测试函数重载 ===");
    console.log(`字符串格式化: ${format("hello")}`);
    console.log(`数字格式化: ${format(1234567)}`);
    console.log(`日期格式化: ${format(new Date())}`);
    console.log(`布尔值格式化: ${format(true)}`);
}

function testDecorator(): void {
    console.log("\n=== 测试装饰器 ===");
    const calc = new Calculator();
    console.log("加法结果:", calc.add(5, 3));
    console.log("乘法结果:", calc.multiply(4, 6));
}

async function testAsyncOperations(): Promise<void> {
    console.log("\n=== 测试异步操作 ===");
    try {
        const response = await fetchData<{ name: string; age: number }>("/api/user");
        console.log("API 响应:", response);
    } catch (error) {
        console.error("API 调用失败:", error);
    }
}

// 主函数
async function main(): Promise<void> {
    console.log("TypeScript 语言练习题");
    console.log("请完成上面标有 TODO 的函数实现");
    console.log("然后运行测试函数查看结果");
    console.log("=====================================");
    
    // 运行测试函数
    testCalculate();
    testStringProcessing();
    testMinMax();
    testStudent();
    testTaskManagement();
    testGenericConstraints();
    testFunctionOverloads();
    testDecorator();
    await testAsyncOperations();
    
    console.log("\n练习完成后，你可以尝试以下进阶挑战：");
    console.log("1. 实现一个类型安全的状态管理器");
    console.log("2. 创建一个支持插件系统的应用框架");
    console.log("3. 使用装饰器实现依赖注入系统");
    console.log("4. 实现一个类型安全的 ORM 查询构建器");
    console.log("5. 创建一个支持类型推断的配置系统");
}

// 运行主函数
main().catch(console.error);

// 导出类型和接口供其他模块使用
export {
    CalculationResult,
    StringOperation,
    MinMax,
    Student,
    StudentClass,
    TaskStatus,
    Task,
    ApiResponse,
    DeepReadonly,
    Pick
};

File: grammar/go/5_exercises.go
package main

import (
	"fmt"
	"math"
	"strings"
)

// ==================== 练习题 ====================
// 请完成以下练习题，每个函数都有详细的说明和示例

// 练习1: 变量和常量
// 完成以下函数，计算圆的面积和周长
// 提示：使用 math.Pi 常量
func calculateCircle(radius float64) (area float64, perimeter float64) {
	// TODO: 实现计算圆的面积和周长
	area = math.Pi * radius * radius
	// perimeter = 2 * π * r
	perimeter = 2 * math.Pi * radius
	return area, perimeter
}

// 练习2: 字符串操作
// 完成函数，将输入的字符串转换为标题格式（每个单词首字母大写）
// 例如："hello world" -> "Hello World"
func toTitleCase(input string) string {
	// TODO: 使用 strings 包的函数来实现
	// 提示：可以使用 strings.Fields() 和 strings.Title() 或 strings.ToTitle()
	words := strings.Fields(input)
	for i, word := range words {
		words[i] = strings.Title(word)
	}
	return strings.Join(words, " ")
}

// 练习3: 切片操作
// 完成函数，找出切片中的最大值和最小值
func findMinMax(numbers []int) (min int, max int) {
	if len(numbers) == 0 {
		return 0, 0
	}
	max = numbers[0]
	min = numbers[0]
	for _, v := range numbers {
		if v >= max {
			max = v
		}
		if v <= min {
			min = v
		}
	}

	return min, max // 请替换为正确的实现
}

// 练习4: 结构体和方法
// 定义一个学生结构体，包含姓名、年龄和成绩切片
// 实现计算平均成绩的方法和添加成绩的方法
type ExerciseStudent struct {
	Name   string
	Age    int
	Grades []float64
}

// 为 ExerciseStudent 实现一个方法，计算平均成绩
func (s ExerciseStudent) AverageGrade() float64 {
	// TODO: 计算并返回平均成绩
	// 如果没有成绩，返回 0
	if len(s.Grades) == 0 {
		return 0.0
	}
	var sum float64 = 0.0
	for _, grade := range s.Grades {
		sum += grade
	}
	return sum / float64(len(s.Grades))
}

// 为 ExerciseStudent 实现一个指针方法，添加新成绩
func (s *ExerciseStudent) AddGrade(grade float64) {
	s.Grades = append(s.Grades, grade)
}

// 练习5: 指针操作
// 完成函数，交换两个整数的值
func swapIntegers(a, b *int) {
	// TODO: 使用指针交换两个整数的值
	*a,*b = *b,*a

}

// 练习6: 错误处理
// 完成函数，安全地进行除法运算
func safeDivide(a, b float64) (result float64, err error) {
	// TODO: 实现安全除法
	// 如果 b 为 0，返回错误
	if b == 0 {
		return 0, fmt.Errorf("除数不能为0")
	}
	// 否则返回 a/b 的结果
	return a/b, nil
}

// 练习7: 接口
// 定义一个 Shape 接口，包含 Area() 和 Perimeter() 方法
// 实现一个 Rectangle 结构体来满足这个接口
type ExerciseShape interface {
	Area() float64
	Perimeter() float64
}

// 矩形结构体
type ExerciseRectangle struct {
	Width  float64
	Height float64
}

// TODO: 为 Rectangle 实现 Shape 接口的方法
func (r ExerciseRectangle) Area() float64 {
	return r.Width * r.Height
}

func (r ExerciseRectangle) Perimeter() float64 {
	return 2 * (r.Width + r.Height)
}

// 练习8: 高阶函数
// 完成函数，对切片中的每个元素应用给定的函数
func mapSlice(numbers []int, fn func(int) int) []int {
	// TODO: 实现 map 函数
	// 对 numbers 中的每个元素应用 fn 函数
	newNumbers := make([]int, len(numbers))
	for i,v := range numbers{
		newNumbers[i] = fn(v)
	}
	// 返回新的切片
	return newNumbers
}

// 练习9: 并发（可选，较难）
// 使用 goroutine 和 channel 计算数字的平方
func calculateSquares(numbers []int) []int {
	// TODO: 使用 goroutine 并发计算每个数字的平方
	// 提示：创建一个 channel 来收集结果
	return nil
}

// ==================== 测试函数 ====================
// 以下是一些测试函数，用于验证你的实现

func testCalculateCircle() {
	fmt.Println("=== 测试圆形计算 ===")
	area, perimeter := calculateCircle(5.0)
	fmt.Printf("半径为5的圆：面积=%.2f, 周长=%.2f\n", area, perimeter)
	// 期望结果：面积≈78.54, 周长≈31.42
}

func testTitleCase() {
	fmt.Println("\n=== 测试标题格式转换 ===")
	result := toTitleCase("hello world go programming")
	fmt.Printf("转换结果: %s\n", result)
	// 期望结果："Hello World Go Programming"
}

func testMinMax() {
	fmt.Println("\n=== 测试最大最小值 ===")
	numbers := []int{3, 7, 2, 9, 1, 5, 8}
	min, max := findMinMax(numbers)
	fmt.Printf("数组 %v 的最小值: %d, 最大值: %d\n", numbers, min, max)
	// 期望结果：最小值=1, 最大值=9
}

func testStudent() {
	fmt.Println("\n=== 测试学生结构体 ===")
	// TODO: 创建学生实例并测试方法
	student := ExerciseStudent{Name: "张三", Age: 20, Grades: []float64{85.5, 92.0, 78.5}}
	fmt.Printf("学生信息: %+v\n", student)
	fmt.Printf("平均成绩: %.2f\n", student.AverageGrade())
	student.AddGrade(88.0)
	fmt.Printf("添加成绩后平均分: %.2f\n", student.AverageGrade())
}

func testSwap() {
	fmt.Println("\n=== 测试交换函数 ===")
	a, b := 10, 20
	fmt.Printf("交换前: a=%d, b=%d\n", a, b)
	swapIntegers(&a, &b)
	fmt.Printf("交换后: a=%d, b=%d\n", a, b)
	// 期望结果：a=20, b=10
}

func testSafeDivide() {
	fmt.Println("\n=== 测试安全除法 ===")
	result, err := safeDivide(10, 2)
	if err != nil {
		fmt.Printf("错误: %v\n", err)
	} else {
		fmt.Printf("10 ÷ 2 = %.2f\n", result)
	}

	result, err = safeDivide(10, 0)
	if err != nil {
		fmt.Printf("错误: %v\n", err)
	} else {
		fmt.Printf("10 ÷ 0 = %.2f\n", result)
	}
}

func testShape() {
	fmt.Println("\n=== 测试形状接口 ===")
	// TODO: 创建矩形实例并测试接口方法
	rect := ExerciseRectangle{Width: 5, Height: 3}
	fmt.Printf("矩形 %+v\n", rect)
	fmt.Printf("面积: %.2f\n", rect.Area())
	fmt.Printf("周长: %.2f\n", rect.Perimeter())
	
	// 接口使用
	var shape ExerciseShape = rect
	fmt.Printf("通过接口访问面积: %.2f\n", shape.Area())
}

func testMapSlice() {
	fmt.Println("\n=== 测试映射函数 ===")
	numbers := []int{1, 2, 3, 4, 5}

	// 测试平方函数
	fmt.Printf("原数组: %v\n", numbers)
	squares := mapSlice(numbers, func(x int) int { return x * x })
	
	fmt.Printf("平方后: %v\n", squares)

	// 测试加倍函数
	doubled := mapSlice(numbers, func(x int) int { return x * 2 })
	fmt.Printf("加倍后: %v\n", doubled)
}

func main5() {
	fmt.Println("Go 语言练习题")
	fmt.Println("请完成上面标有 TODO 的函数实现")
	fmt.Println("然后运行测试函数查看结果")
	fmt.Println("=====================================")

	// 运行测试函数
	testCalculateCircle()
	testTitleCase()
	testMinMax()
	testStudent()
	testSwap()
	testSafeDivide()
	testShape()
	testMapSlice()

	fmt.Println("\n练习完成后，你可以尝试以下进阶挑战：")
	fmt.Println("1. 实现一个简单的计算器")
	fmt.Println("2. 创建一个学生管理系统")
	fmt.Println("3. 使用 goroutine 实现并发计算")
	fmt.Println("4. 实现自定义排序算法")
}
